CMakeLists.txt

```diff
@@ -0,0 +1,511 @@
+
+	CHECK_TYPE_SIZE (pthread_rwlock_t UPNP_USE_RWLOCK)
+	UNSET (CMAKE_EXTRA_INCLUDE_FILES)
+	UNSET (CMAKE_REQUIRED_INCLUDES)
+ENDIF (TARGET Threads::Threads)
+
+IF (TARGET Threads::Threads AND NOT TARGET Threads::Static)
+	SET_TARGET_PROPERTIES (Threads::Threads PROPERTIES
+		IMPORTED_GLOBAL TRUE
+	)
+
+	ADD_LIBRARY (Threads::Static ALIAS Threads::Threads)
+ENDIF (TARGET Threads::Threads AND NOT TARGET Threads::Static)
+
+IF (open_ssl)
+	INCLUDE (FindOpenSSL)
+
+	IF (NOT OPENSSL_FOUND)
+		MESSAGE (FATAL_ERROR "openssl not found")
+	ENDIF (NOT OPENSSL_FOUND)
+ENDIF (open_ssl)
+
+CONFIGURE_FILE (${UPNP_SOURCE_DIR}/upnp/inc/upnpconfig.h.cm ${UPNP_BINARY_DIR}/upnp/inc/upnpconfig.h)
+CONFIGURE_FILE (${UPNP_SOURCE_DIR}/upnp/sample/common/config_sample.h.cm ${UPNP_BINARY_DIR}/upnp/sample/common/config_sample.h)
+CONFIGURE_FILE (${UPNP_BINARY_DIR}/autoconfig.h.cm ${UPNP_BINARY_DIR}/autoconfig.h)
+
+ADD_SUBDIRECTORY (ixml)
+ADD_SUBDIRECTORY (upnp)
+
+ENABLE_TESTING()
+
+FILE (GLOB XML_FILES
+	${CMAKE_CURRENT_SOURCE_DIR}/ixml/test/testdata/*.xml
+)
+
+IF (NOT TARGET check)
+	ADD_CUSTOM_TARGET (check
+		COMMAND ${CMAKE_CTEST_COMMAND}
+	)
+ENDIF (NOT TARGET check)
+
+ADD_DEPENDENCIES (check
+	test_ixml_shared
+	test_ixml_static
+	test_upnp_init
+	test_upnp_init_static
+	test_upnp_list
+	test_upnp_list_static
+	test_upnp_log
+	test_upnp_log_static
+	test_upnp_url
+	test_upnp_url_static
+)
+
+ADD_TEST (NAME TEST_IXML_SHARED
+	COMMAND test_ixml_shared ${XML_FILES}
+)
+
+ADD_TEST (NAME TEST_IXML_STATIC
+	COMMAND test_ixml_static ${XML_FILES}
+)
+
+ADD_TEST (NAME TEST_UPNP_INIT
+	COMMAND test_upnp_init
+)
+
+ADD_TEST (NAME TEST_UPNP_INIT_STATIC
+	COMMAND test_upnp_init_static
+)
+
+ADD_TEST (NAME TEST_UPNP_LIST
+	COMMAND test_upnp_list
+)
+
+ADD_TEST (NAME TEST_UPNP_LIST_STATIC
+	COMMAND test_upnp_list_static
+)
+
+ADD_TEST (NAME TEST_UPNP_LOG
+	COMMAND test_upnp_log
+)
+
+ADD_TEST (NAME TEST_UPNP_LOG_STATIC
+	COMMAND test_upnp_log_static
+)
+
+ADD_TEST (NAME TEST_UPNP_URL
+	COMMAND test_upnp_url
+)
+
+ADD_TEST (NAME TEST_UPNP_URL_STATIC
+	COMMAND test_upnp_url_static
+)
+
+ADD_TEST (NAME TEST_UPNP_UNITTEST
+	COMMAND test_upnp_unittest
+)
+
+ADD_TEST (NAME TEST_UPNP_UNITTEST_STATIC
+	COMMAND test_upnp_unittest_static
+)
+
+INSTALL (EXPORT UPNPCONFIG
+	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/UPNP
+)
+
+INCLUDE (CMakePackageConfigHelpers)
+
+CONFIGURE_PACKAGE_CONFIG_FILE (
+	UPNP.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/UPNP.cmake
+	INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/UPNP
+)
+
+WRITE_BASIC_PACKAGE_VERSION_FILE (UPNPVersion.cmake
+	VERSION ${UPNP_VERSION}
+	COMPATIBILITY SameMajorVersion
+)
+
+INSTALL (FILES
+	${CMAKE_CURRENT_BINARY_DIR}/UPNP.cmake
+	${CMAKE_CURRENT_BINARY_DIR}/UPNPVersion.cmake
+	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/UPNP
+)
+
+SET (VERSION ${UPNP_VERSION_STRING})
+SET (prefix ${CMAKE_INSTALL_PREFIX})
+SET (exec_prefix "\${prefix}")
+SET (libdir "\${exec_prefix}/${CMAKE_INSTALL_LIBDIR}")
```

Comment:
Note that `CMAKE_INSTALL_*` variables might be absolute paths, in which case the pkg-config will contain wrong paths. See also https://github.com/jtojnar/cmake-snips#concatenating-paths-when-building-pkg-config-files

Reply:
Good to know. I'll take a look at this in the next week.